### ISSUE TITLE
how to save the trained model for inference?

### ISSUE THREAD

**Issue by seyeeet** [facebookresearch/faiss#2087]
# Summary

I am using the following sample to train a kmean, it works perfectly fine.
My question is how I can use the train model without retraining it over and over for inference time?
is there a way to save it somewhere and reload it for testing? I checked the tutorial but could not get any answer for my question.

``` python

def run_kmeans(x, nmb_clusters,gpu_ids=None, verbose=True):
    
    """Runs kmeans on 1 GPU.
    Args:
        x: data
        nmb_clusters (int): number of clusters
    Returns:
        list: ids of data in each cluster
    """
    n_data, d = x.shape

    print('FAISS: using GPUs {}'.format(gpu_ids))
    print('We have {} samples and dimention is for each is {}'.format(n_data,d))

    # faiss implementation of k-means
    clus = faiss.Clustering(d, nmb_clusters)

    # Change faiss seed at each k-means so that the randomly picked
    # initialization centroids do not correspond to the same feature ids
    # from an epoch to another.
    clus.seed = np.random.randint(1234)

    clus.niter = 20
    # nredo.nredo = 10
    clus.max_points_per_centroid = 10000000
    res = faiss.StandardGpuResources()
    flat_config = faiss.GpuIndexFlatConfig()
    flat_config.useFloat16 = False
    flat_config.device = gpu_ids
    index = faiss.GpuIndexFlatL2(res, d, flat_config)

    # perform the training
    clus.train(x, index)

    dists, pred_labels = index.search(x, 1) # # we want to see 1 nearest neighbors
    # pred_labels = pred_labels.squeeze()

    # nmi = normalized_mutual_info_score(labels, pred_labels)

    # print("NMI: {}".format(nmi))
    stats = clus.iteration_stats
    losses = np.array([
        stats.at(i).obj for i in range(stats.size())
    ])
    if verbose:
        print('k-means loss evolution: {0}'.format(losses))

    return [int(n[0]) for n in pred_labels], losses[-1] 

```

Running on:
- [ ] CPU
- [x ] GPU

Interface: 
- [ ] C++
- [x ] Python

**Comment by mdouze** [comment]
You can save the centroids, stored in `clus.centroids`, then add them back to the GPU index object.

**Comment by seyeeet** [comment]
Thank you for your help. can you show me in a small snippet how it would be possible please?

**Comment by mdouze** [comment]
`np.save("x", clus.centrdoids)`

**Comment by vinit13792** [comment]
I didn't understand this code, is there a sample snippet where I can check how to run my faiss model on inference mode for deployment?

**Comment by rohitdwivedula** [comment]
@vinit13792 I assume you want to: (1) save a trained KMeans model to file (by saving the centroids), and (2) load the model later and assign clusters to data points.

For step 1, you do something that looks like this: 

```python3
kmeans = faiss.Kmeans(inpuit_dim, num_clusters, verbose=True, gpu=True, niter=300, nredo=10, seed=42)
kmeans.train(knn_data)

# save centroids to file
with open("centroids.npy", 'wb') as f:
        np.save(f, kmeans.centroids)
```

To load the centroids and use them, do: 
```python3
centroids = np.load(centroids.npy")
kmeans = faiss.Kmeans(inpuit_dim, num_clusters, verbose=True, gpu=True, niter=0, nredo=0, seed=42)
kmeans.train(logits, init_centroids=centroids) # this ensures that kmeans.index is created
assert np.sum(kmeans.centroids - centroids) == 0, "centroids are not the same" # sanity check
cluster_distances, cluster_indices = kmeans.assign(new_data)
```

**Comment by sahyagiri** [comment]
From where do we get "logits" variable initialized?

**Comment by joangog** [comment]
> ```python
> assert np.sum(kmeans.centroids - centroids) == 0, "centroids are not the same"
> ```

Doesn't that force us to retrain again?
